### PR TITLE
Restructurizing for easy TS migration

### PR DESCRIPTION
**Pull request by michalusio**
TL;DR:

This PR moves all the source files into an `src` directory and sets up TypeScript so building the project creates the `dist` directory with all JS, D.TS and D.TS.MAP as they should be.
That enables us to add `src` and `example` folders into `.npmignore` to keep the package size smaller.

- Moved source files into `src` directory
- Changed tsconfig.json to transpile `*.js` files (so we can separate `src` from `dist`)
- Moved `typescript` to devDependencies (so users of the package do not need to install TS themselves)
- Added `@types/node` for node 8 (provides better typing for node internals)
- Added `dist` to gitignored (they're regenerated on every build)
- Added `src` to npmignored (don't need to publish it, `dist` has everything)
- Added `examples` to npmignored (don't need to publish it, examples can stay in the repo)
- Added `prepublishOnly` hook (so TS builds everything before publishing a new version)

### CODE REVIEW

**Comment by kenan238** [comment]
Checking this out
